Add configuration file for node initialisation

phklive opened the pull request:
In this PR I propose an addition of a configuration file for the initialisation of the Miden-node.

This PR will: 
- Remove the hard coding of genesis values
- Enable instantiation of the node using a `genesis.toml` file
- Enable the creation of a genesis.dat file containing the serialised GenesisState
- Enable the creation of a file for each created account containing account info + keypair
- Add new structs for managing serialisation and deserialisation of types in Miden ( could be re-used in the client for interop )

Closes: #127 

**Current status:**
✅ The `genesis.toml` configuration file is correctly parsed
✅ The parsed data is correctly deserialized into an `GenesisInput` struct
✅ (`Account`, `seed`) and `Keypair` are correctly generated from `GenesisInput` `accounts` field
✅ Using Vec<Account> and `GenesisInput` data a `GenesisState` is correctly instantiated
✅ Generate `genesis.dat` file
✅ Make sure generated files have the right format / extension (`.mac` "Miden Account File")
✅ Generate account files containing the correct data in the correct format (sync with @igamigo)

**Next steps:**
❌ Add `notes` and `nullifiers` to `genesis.toml` and `GenesisInput` to pre-populate the DB

phklive commented:
Next steps: 
- Generate account files containing the correct information in the correct format
- Make sure we handle the passing of keypairs and associated data to the user correctly
- Generate `genesis.dat` file
- Make sure file have the right format / extension for the different generated files
- Add `notes` and `nullifiers` to `genesis.toml` and `GenesisInput` to pre-populate the notes and nullifiers DB in the store
- Clean PR's and make sure each structs and functions are located in the right file for correct code organisation

bobbinth commented on genesis.toml:
```diff
@@ -0,0 +1,19 @@
+[[accounts]]
+type = "BasicWallet"
+mode = "RegularAccountImmutableCode"
+account_seed = "a123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
+auth_seed = "b123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
+auth_scheme = "RpoFalcon512"
```
A couple of comments:

- I'm not sure we should use `account_seed` here as `account_seed` has a slightly different meaning (i.e., it is a seed which we get after we create the account). The term we usually use for this is `init_seed` - but I think here it might be confusing. Maybe we just say `seed`.
- I would probably move `auth_seed` after `auth_scheme`.
- We usually prefix `hex` data with `0x` to make it clear that this is hex encoding.

phklive replied:
fixed

bobbinth commented on node/Cargo.toml:
```diff
@@ -27,6 +27,7 @@ serde = { version = "1.0", features = ["derive"] }
 tokio = { version = "1.29", features = ["rt-multi-thread", "net", "macros"] }
 tracing = { workspace = true }
 tracing-subscriber = { workspace = true }
+hex = { version = "0.4" }
```
minor nit: let's maintain alphabetical order here.

bobbinth commented on store/src/genesis.rs:
```diff
@@ -1,16 +1,55 @@
+    pub max_supply: u64,
+}
+
+#[derive(Debug, Deserialize)]
 pub struct GenesisState {
```
Why do we need `Deserialize` here?

bobbinth commented on genesis.toml:
I would move this file under `node` directory.

bobbinth commented on node/src/main.rs:
```diff
@@ -4,8 +4,8 @@ use clap::{Parser, Subcommand};
 mod config;
-pub(crate) mod genesis;
 
+const INPUT_GENESIS_FILE_PATH: &str = "genesis.toml";
```
minor nit: I would rename this to `GENESIS_INPUT_FILE_PATH`.

bobbinth commented on store/src/genesis.rs:
I would move all the `*Input` structs into the `node` crate.

bobbinth commented on node/src/commands.rs:
Rather than deleting `genesis.rs` file - I'd move most of the new code (and maybe some existing code) from here to the `genesis.rs` file. This file can also be the place where we define `*Input` structs which are now in the `store` create.

This will nicely encapsulate the logic for generating `genesis.dat` file in one place.

bobbinth reviewed (commented):
Thank you! Looks good! I left some preliminary comments inline.

phklive commented:
Work in the latest commit: 
- Removed `async` for `make_genesis` function that did not use it
- Added testing for the `make_genesis` function
- Refactored files by: removing `commands.rs`, removing `config.rs`, adding `start.rs`. Now the logic for config and start are encapsulated in the `start.rs` file, considering that the config is needed for the `start` function it's logic for them to be encapsulated together. The genesis functions live in the `genesis.rs` file, and the `commands.file` was then empty hence removed.

phklive commented:
In the latest commit I am able to generate an accounts folder containing all account files named `account[n].mac`.

They are serialised using serde_json, outputting the following code:

```json
{
  "account": [
    98,
    102,
    209,
    221,
    42,
    33,
    234,
    84,
    0,
    0,
    0,
    0,
    ...
  ],
  "seed": [
    10766778793097982000,
    6665883180922515000,
    9172510378277381000,
    10594060660920822000
  ],
  "auth_scheme": "RpoFalcon512",
  "auth_seed": "0xb123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef0123456789abcdef"
}
```

They are serialised using the following struct: 

```rust
#[derive(Debug, Serialize, Deserialize)]
pub struct AccountData {
    pub account: Account,
    pub seed: Word,
    pub auth_scheme: String,
    pub auth_seed: String,
}
```

If validated I will open a PR on miden base to add this transport struct next to `Account` enabling use in all miden-crates.

To recap: 
- We parse a `genesis.toml` config file containing `version`, `timestamp` and `accounts`
- We create a `genesis.dat` file that is usable to run the node containing a `GenesisState`
- We create an `accounts` folder containing all accounts files serialised using serde_json

igamigo commented:
I think the toml plus one-file-per-account file looks good. 
If we were to add a struct such as the `AccountData` struct you described, however, I would try to have stronger typed fields in it such as `[u8;32]` for the seed and an enum for the `auth_scheme`.

bobbinth reviewed (changes requested):
Thank you! Looks good! I left some comments inline.

A few additional comments:

Now looking at this, I probably wouldn't serialize the file as JSON. It's just a bunch of binary data - so, might as well serialize it as binary. Let's use our `Serializable` and `Deserializable` traits for that.

Regarding `AccountData` struct. I'd probably do it as follows:
```Rust
pub struct AccountData {
    pub account: Account,
    pub seed: Option<Word>,
    pub auth_info: AuthInfo,
}

pub enum AuthInfo {
    RpoFalcon512Seed([u8; 40]),
}
```
A couple of reasons for the above:
- The `seed` is optional to support the scenario for when we may want to import an existing (rather than a new) account. This is not relevant here, but if we are developing a generic format that the client can import - this would be good to support.
- It would be good to add a bit more structure to authentication info.

Lastly, I don't love the `AccountData` name, but at the moment don't have a better suggestion.

phklive commented on node/genesis-example.toml:
I tried to follow the same convention as in the other folders, for example: 
- https://github.com/0xPolygonMiden/miden-node/blob/main/rpc/rpc-example.toml
- https://github.com/0xPolygonMiden/miden-node/blob/main/store/store-example.toml
- https://github.com/0xPolygonMiden/miden-node/blob/main/block-producer/block-producer-example.toml

What do you think? Should we all standardize them and remove the `-example` and add a comment at the top to say that it's an example config or should we keep the example for all `-example` config files. 

Or something else?

bobbinth replied:
I don't think that for component executables `miden.toml` is the default config file (though, I might be wrong). If so, i'd just keep them as is.

We'd also probably refactor how we start individual components when addressing #92.

phklive replied:
I did not get what you meant / requested here. 

I believe that `miden.toml` is a default configuration file used when we run `cargo run start` on the node. 

Hence we can consider it as an example imo.

bobbinth replied:
Right, `miden.toml` is the default configuration file for the node, but there is no default configuration file for the individual components (rpc, store, block producer) - so, I think it is fine to keep `example_` prefix for individual components, but I'd keep it as `miden.toml` for the node.

phklive commented on store/src/genesis.rs:
```diff
@@ -9,8 +9,7 @@ use miden_objects::{
 use crate::state::ACCOUNT_DB_DEPTH;
 
 pub const GENESIS_BLOCK_NUM: u32 = 0;
-
-/// Represents the state at genesis, which will be used to derive the genesis block.
```
Typo

phklive commented on node/src/start.rs:
What should be in the `config.rs` file? The previous config and keep only the start function in start? Something else?

bobbinth replied:
I was thinking things like `NodeTopLevelConfig`, `CONFIG_FILENAME` etc. - but I guess those probably belong to their individual files. So, we can skip `config.rs`.

Btw, should we rename `NodeTopLevelConfig` to `StartCommandConfig`?

phklive replied:
Yes agreed

phklive commented on node/src/genesis.rs:
```diff
@@ -1,26 +1,298 @@
+    })?;
+
+    Ok(())
+}
```
Here, why does `new` return a Result ? 

when you are saying `write()` are you referring to `write_into()` as used in serialisation here: 
https://github.com/0xPolygonMiden/miden-base/blob/1841b99c776722854444669cc2168e2eed49bac2/objects/src/accounts/mod.rs#L173

Or 

Directly as write in "write to a file" as done here with:
```rust
fs::write(file_path.as_path(), json_data).map_err(|err| {
        anyhow!("Failed to write account file to {}, Error: {err}", file_path.display())
    })?;
```

bobbinth replied:
> Here, why does `new` return a Result ?

I was thinking to cover the case of when parsing of auth info (scheme and seed) fails.

For the `write()` function, I was thinking that it would be the second option (write to file).

phklive commented on node/src/genesis.rs:
```diff
@@ -1,26 +1,298 @@
+                let seed: [u8; 32] = hex_string_to_byte_array(&inputs.seed)?;
+                let auth_seed: [u8; 40] = hex_string_to_byte_array(&inputs.auth_seed)?;
+
+                let keypair = KeyPair::from_seed(&auth_seed)?;
```
I tried not to create a complex data type for Auth because I am not sure where it would live. 

Would we add it to miden_lib in this file: 
https://github.com/0xPolygonMiden/miden-base/blob/1841b99c776722854444669cc2168e2eed49bac2/miden-lib/src/auth.rs#L4-L10

phklive replied:
And it seems that it would add many new types, also not considering the *input ones: 
- AuthScheme (already exists)
- AuthInput
- AuthInfo

Should we find a better way to handle this or does that seem okay for you?

bobbinth replied:
Yep - we may need to find a better way, but I'd probably leave this normalization for the PR when we move these structs into `miden-base`.

phklive commented:
In the latest commit added: 
- Serialization and Deserialization of `AccountData`
- implementation of `write` and `read` on `AcccountData` enabling writing and reading `AccountData` from files
- Tests to verify the correct functionality of the serialization and deserialization 

While implementing the tests I face a blocker and wanted to know it you had advice regarding this information: 

My test workflow is the following: 
1. I generate the `genesis.toml` file
2. I run `make_genesis`
3 Verify that `genesis.dat` and 2 account files have been created
4. Deserialise 2 `AccountData` from account files
5. Create 2 `Account` from known data (what it should be)
6. Create 2 new `AccountData` using the 2 accounts
7. Compare the 2 deserialized `AccountData` vs 2 created `AccountData` (what they should be)

What blocks me is that the tests are not stable, where they sometimes pass and sometimes don't as if there were a part of randomness in the process that I do not grasp yet. 

I have printed out the output of the `assert_eq!()` macro and when analysing it for the `basic_wallet` `AccountData` everything is the same except those:  
```
AccountCode (ModuleAst and Related Fields)

Local Procedures, Reexported Procedures: The procedures listed are the same in both left and right.

Import Info:

Left: "basic": "miden::contracts::wallets::basic"
Right: "basic": "miden::contracts::auth::basic", "basic_wallet": "miden::contracts::wallets::basic"

Procedures and Procedure Tree: Both sides appear to have the same procedure digests but differ in the procedure_tree:

Left: OnceCell(<uninit>)
Right: OnceCell(SimpleSmt { ... })

Differences exist in the import_info and procedure_tree.
```

We can see it here: 
![SCR-20240112-reri](https://github.com/0xPolygonMiden/miden-node/assets/42912740/8a7c381f-d318-4b94-bf4a-71cc6682618a)
![SCR-20240112-rfby](https://github.com/0xPolygonMiden/miden-node/assets/42912740/043599b0-f0d8-4570-979b-c00272f64420)

This makes the test fail. Would someone have advice? @bobbinth or other members

Once this function works: 
- Cleaning of PR
- Improvement of code organisation with `AccountData` and related structs that will be added to miden-base
- Addition of logic of account import in the miden-client: https://github.com/0xPolygonMiden/miden-client/issues/86
- Addition of logic for management and import of notes and nullifiers

bobbinth commented:
> What blocks me is that the tests are not stable, where they sometimes pass and sometimes don't as if there were a part of randomness in the process that I do not grasp yet.

This could be a part of the account seed generation process. If `miden-lib` is imported with the `concurrent` feature enabled (which I think is the case for `miden-node`), the seed generator will use multiple threads to find the account seed. This process is non-deterministic and so could give different results at different times.

This would not explain, however, the differences in account code. These may be attributed to serializing with different attributes enabled (this is described in https://github.com/0xPolygonMiden/miden-vm/issues/1177).

phklive commented:
Continuing the search I can confirm that all the inputs passed to both the `create_basic_wallet()` in the tested `make_genesis()` function or in the manually created `AccountData` are exactly the same. Hence normally deterministically generating the same account. 

Nonetheless printing out the `AccountType` in the `make_genesis()` function I go the following warning message:
![SCR-20240115-kmsf](https://github.com/0xPolygonMiden/miden-node/assets/42912740/03fb96e4-8714-4717-8f98-3e2fbbf08251)

Seems to explain why I have the `auth` in the manually generated one but not in the `make_genesis()` created one.

phklive commented:
> > What blocks me is that the tests are not stable, where they sometimes pass and sometimes don't as if there were a part of randomness in the process that I do not grasp yet.
> 
> This could be a part of the account seed generation process. If `miden-lib` is imported with the `concurrent` feature enabled (which I think is the case for `miden-node`), the seed generator will use multiple threads to find the account seed. This process is non-deterministic and so could give different results at different times.
> 
> This would not explain, however, the differences in account code. These may be attributed to serializing with different attributes enabled (this is described in [0xPolygonMiden/miden-vm#1177](https://github.com/0xPolygonMiden/miden-vm/issues/1177)).

The `AstSerdeOptions` is a `const` defined as `true`: 
https://github.com/0xPolygonMiden/miden-base/blob/f792b0ce6d7323491d357c8d74f58f8359f83ae5/objects/src/accounts/code.rs#L11-L15

Hence I don't think that my implementation would be changing the serialisation settings.

igamigo commented:
So for this:

> Procedures and Procedure Tree: Both sides appear to have the same procedure digests but differ in the procedure_tree:
>
> Left: OnceCell(<uninit>)
> Right: OnceCell(SimpleSmt { ... })

It seems to be that the `OnceCell` is not initialized for the retrieved object. You should be able to do `let _ = account.code().procedure_tree();` before comparing and have the object be the same as the other.

bobbinth commented:
A few thoughts here:

- The warnings printed out by `miden-vm` don't seem correct to me. I think both imports are needed by the basic account contract. But I haven't looked too deeply into this.
- Comparing `AccountCode`'s directly should work fine as [equality checking](https://github.com/0xPolygonMiden/miden-base/blob/main/objects/src/accounts/code.rs#L147) is implemented manually on `AccountCode` - so, it should ignore the `OnceCell` part.

Is it correct to say that basically everything is the same for both deserialized and generated accounts except for `account.code.module` field? In for that field, the difference is in imports only?

phklive commented:
Regarding the discussion of domain and wrapper types that we had yesterday I have the feeling that the *Input types of this PR should be located elsewhere. Would be glad to have your POV on this situation @bobbinth , @hackaugusto , @plafer.

https://github.com/0xPolygonMiden/miden-node/pull/138/files#diff-ec1a6016d771bb9c9a4523a0310eab06db32674af01cd54859c59c27fadb1784R26-R68

igamigo commented:
I think the readme could be modified with updated instructions and account for some of these changes

bobbinth commented on node/genesis.toml:
```diff
@@ -0,0 +1,21 @@
+# This is an example configuration file for the Miden node.
```
We should update this text - i.e., this is an example genesis input file rather than a configuration file for the ndoe.

bobbinth reviewed (changes requested):
Thank you! I pushed a commit which refactored the code somewhat and also left additional comments inline.

Also, as @igamigo suggested with should update README documentation to reflect these changes.

bobbinth commented on node/src/main.rs:
```diff
@@ -19,12 +26,17 @@ pub struct Cli {
     /// Start the node
     Start {
-        #[arg(short, long, value_name = "FILE", default_value = config::CONFIG_FILENAME)]
+        #[arg(short, long, value_name = "FILE", default_value = NODE_CONFIG_FILE_PATH)]
         config: PathBuf,
     },
 
     /// Generate genesis file
     MakeGenesis {
+        /// Read genesis file inputs from this location
+        #[arg(short, long, value_name = "FILE", default_value = DEFAULT_GENESIS_INPUTS_PATH)]
```
Is using `value_name = "FILE"` here for both commands correct? When I execute `miden-node make-genesis --help`, I get the following:
```
Generate genesis file

Usage: miden-node make-genesis [OPTIONS]

Options:
  -i, --input-path <FILE>          Read genesis file inputs from this location [default: genesis.toml]
  -o, --output-path <OUTPUT_PATH>  Write the genesis file to this location [default: genesis.dat]
  -f, --force                      Generate the output file even if a file already exists
  -h, --help                       Print help
```

phklive replied:
Are you saying that you would prefer to have <CONFIG> and <INPUT_PATH> instead. Or should we put something else?

bobbinth replied:
I'm actually not sure what `value_name` is supposed to represent. Having `output-path <OUTPUT_PATH>` or `input-path <INPUT_PATH>` is kind of redundant, I think. Maybe we don't need it at all?

phklive replied:
Having not build the CLI myself I gave a deeper look into the Clap crate, it is seems that Clap does not provide a standard way to not display value_name for commands that take options.

Hence we could keep it as <FILE> or something else. Or find a way to hide it ( could be time-consuming ). 

I found an example with another instantiation of that seemed to remove the value_name, but that would need for me to make a PR related to changing the CLI structure.

bobbinth replied:
Yeah - no need to do anything custom here. What happens if we just remove `value_name` from the attribute? I.e., do something like this:
```Rust
MakeGenesis {
    /// Read genesis file inputs from this location
    #[arg(short, long, default_value = DEFAULT_GENESIS_INPUTS_PATH)]
    inputs_path: PathBuf,

    /// Write the genesis file to this location
    #[arg(short, long, default_value = DEFAULT_GENESIS_FILE_PATH)]
    output_path: PathBuf,

    /// Generate the output file even if a file already exists
    #[arg(short, long)]
    force: bool,
},
```

phklive commented:
Since the last PR the accounts folder is now added alongside the genesis file.

mFragaBA commented:
should we add to the `.gitignore` the `accounts` directory?

phklive commented:
> should we add to the `.gitignore` the `accounts` directory?

The default would be to create the accounts/ folder in the current_dir, but it can be overridden by the user through the `-o` flag. 

Hence I propose not adding it considering that it is variable. Like the genesis file.

bobbinth commented on node/genesis.toml:
```diff
@@ -0,0 +1,21 @@
+
+[[accounts]]
+type = "BasicWallet"
+mode = "RegularAccountImmutableCode"
```
Let's get rid of the `mode` here and for `BasicFungibleFaucet` as I mentioned in the previous review.

bobbinth commented on node/src/main.rs:
```diff
@@ -19,13 +26,23 @@ pub struct Cli {
-    /// Generate genesis file
+    /// Generates a genesis file and associated account files based on a specified genesis input
+    ///
+    /// This function creates a new genesis file and associated account files at the specified output
+    /// paths. It checks for the existence of the output file, and if it already exists, an error is
+    /// thrown unless the `force` flag is set to overwrite it. The function also verifies the existence
+    /// of a configuration file required for initializing the genesis file.
```
Here, I would say "This command" rather than "This function".

Also, we should probably update the last sentence as we done verify the existence of configuration file here.

bobbinth commented on node/src/commands/genesis/mod.rs:
```diff
@@ -0,0 +1,270 @@
+    let mut accounts_path = PathBuf::from(&parent_path);
+    accounts_path.push(DEFAULT_ACCOUNTS_DIR);
+
+    let accounts_exists = accounts_path.try_exists()?;
+
+    if accounts_exists {
+        if *force {
+            fs::remove_dir_all(&accounts_path)?;
+        } else {
+            return Err(anyhow!("Failed to generate accounts folder {} because it already exists. Use the --force flag to overwrite.", accounts_path.display()));
+        }
+    }
+
+    fs::create_dir_all(&accounts_path)
+        .map_err(|err| anyhow!("Failed to create accounts directory: {err}"))?;
```
Another option to do this would be:

1. Do not delete the directory here.
2. If the directory does not exist, create it.
3. When writing individual account files (on line 169 below), throw an error if a file exists and `force` flag is not on.

The benefit of this is that we never delete the whole directory (which may contain files we are not planning to overwrite).

bobbinth reviewed (approved):
Looks good! Thank you! I left a couple of comments inline. Once these are addressed, we are good to merge.

phklive commented:
@bobbinth If we remove the `value_name` from the attribute it defaults to the name of the option.

![SCR-20240119-jryw](https://github.com/0xPolygonMiden/miden-node/assets/42912740/e6ef5f21-5323-4a52-99e0-60e76fe461da)

![SCR-20240119-jsaa](https://github.com/0xPolygonMiden/miden-node/assets/42912740/922211e7-997a-4337-bd23-f0248b09df98)

I propose finding a meaningful name, FILE could be okay I guess.

bobbinth commented:
> If we remove the `value_name` from the attribute it defaults to the name of the option.

I think that's better since we can't have `FILE` for both options.

phklive commented:
@bobbinth, what is your opinion on these warnings coming from the VM? 

![SCR-20240119-jvle](https://github.com/0xPolygonMiden/miden-node/assets/42912740/2ae31370-9ab9-4b00-b38a-d45e46990913)

Should we open an issue?

bobbinth commented:
> @bobbinth, what is your opinion on these warnings coming from the VM?

I've asked @Overcastan to investigate.

phklive commented:
> > If we remove the `value_name` from the attribute it defaults to the name of the option.
> 
> I think that's better since we can't have `FILE` for both options.

Why wouldn't we be able to have `FILE` for both options? I believe that I didn't understand what you meant.

bobbinth commented:
> Why wouldn't we be able to have `FILE` for both options? I believe that I didn't understand what you meant.

Maybe I tested it wrong, but when we had `FILE` in both places it looked like I pasted in https://github.com/0xPolygonMiden/miden-node/pull/138#discussion_r1457060987.

phklive commented:
> #138 (comment)

![SCR-20240119-kbxs](https://github.com/0xPolygonMiden/miden-node/assets/42912740/f0e2b2aa-9039-429a-b132-12787d11f3ba)

I think that it works, we can have both as "FILE"

bobbinth commented:
> I think that it works, we can have both as "FILE"

Ah! Then yes, let's use `FILE` for both.

Overcastan commented:
> > @bobbinth, what is your opinion on these warnings coming from the VM?
> 
> I've asked @Overcastan to investigate.

The problem is that I added one excess check for unused imports -- I'm testing both programs and modules instead of checking only programs. I'll create a PR today

bobbinth commented:
> The problem is that I added one excess check for unused imports -- I'm testing both programs and modules instead of checking only programs. I'll create a PR today

I think we should check both programs and modules - so, I think that part is fine.

The problem is that we don't know which file the unused imports are detected (e.g., see https://github.com/0xPolygonMiden/miden-node/pull/138#issuecomment-1900080196).

bobbinth commented:
> The problem is that we don't know which file the unused imports are detected (e.g., see [#138 (comment)](https://github.com/0xPolygonMiden/miden-node/pull/138#issuecomment-1900080196)).

Or maybe the message tells us that there are unused imports in these modules? In this case, we should make a small adjustment to the message and instead of say `unused imports: foo::bar` we should say `unused import in foo::bar`.

phklive commented:
@bobbinth could you explain your POV about note and nullifier importation at genesis. 
- How should the notes and nullifiers be read ( format ).
- After deserialisation they should then be sent to the store for storage in the db through rpc calls.
- Other information you have in mind?

Opened an issue: #154

phklive commented:
Updated the Readme. 

Should we wait for another review before merging or should we merge now?

bobbinth commented on README.md:
```diff
@@ -53,16 +53,17 @@ Currently, the only difference between the two is how long the `make-genesis` co
 
-This will generate 3 files in the current directory: 
+By default this will generate 1 file and 1 folder in the current directory: 
 - `genesis.dat`: the genesis file.
-- `faucet.fsk` and `wallet.fsk`: the public/private keys of the faucet and wallet accounts, respectively.
+- `accounts/` folder containing an `account{i}.mac` file for each provided account in the genesis inputs file: a serialised `AccountData` struct containing the account, account-seed, auth-scheme and auth-seed. Learn more: https://github.com/0xPolygonMiden/miden-base/blob/341f692edcb375c0473d8a8c19e2c5944c7c9b57/objects/src/accounts/data.rs#L17-L30
```
minor nit: I'd use "directory" instead of "folder".

For accounts, I probably wouldn't make a reference to the code. Instead we could say something like: 

`accounts` directory containing `.mac` files (one per account) for the accounts defined in the genesis inputs file. Each `.mac` file contains full serialization of an account, including code, storage, and authentication info.

bobbinth commented on node/src/commands/genesis/mod.rs:
```diff
@@ -0,0 +1,272 @@
+    let mut accounts_path = PathBuf::from(&parent_path);
+    accounts_path.push(DEFAULT_ACCOUNTS_DIR);
+
+    let accounts_exists = accounts_path.try_exists()?;
+
+    if !accounts_exists {
```
I would probably combine these into the same line like so:
```Rust
if !accounts_path.try_exists()? {
```

bobbinth reviewed (approved):
Looks good! Thank you! Just a couple more minor nits inline, and let's merge.
